Swagger plugin generator doesn't respect project references

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Current behavior

When you have some packages with project references (composite: true) and then import them to our NestJS app (with tsconfig references option), run generate-metadata.ts, you will get following error
```typescript
error TS6305: Output file 'libs/data-access/dist/index.d.ts' has not been built from source file 'libs/data-access/src/index.ts'.
```

### Minimum reproduction code

...

### Steps to reproduce

_No response_

### Expected behavior

It should respect project references and successfully generate metadata.

### Package version

11.2.3

### NestJS version

11.1.9

### Node.js version

22.18.0

### In which operating systems have you tested?

- [ ] macOS
- [ ] Windows
- [x] Linux

### Other

_No response_

--- a/lib/plugin/visitors/readonly.visitor.ts
+++ b/lib/plugin/visitors/readonly.visitor.ts
@@ -4,10 +4,86 @@ import { isFilenameMatched } from '../utils/is-filename-matched.util';
 import { ControllerClassVisitor } from './controller-class.visitor';
 import { ModelClassVisitor } from './model-class.visitor';
 
+/**
+ * Collects source file names from all transitively referenced TypeScript projects.
+ * Used internally by {@link ReadonlyVisitor.createTsProgram} to avoid TS6305 errors
+ * when a project uses composite project references that have not yet been built.
+ */
+function collectProjectReferenceSourceFiles(
+  projectReferences: readonly ts.ProjectReference[] | undefined,
+  visitedProjects = new Set<string>()
+): string[] {
+  if (!projectReferences) {
+    return [];
+  }
+  const sourceFiles: string[] = [];
+  for (const ref of projectReferences) {
+    const refConfigPath = ts.resolveProjectReferencePath(ref);
+    if (visitedProjects.has(refConfigPath)) {
+      continue;
+    }
+    visitedProjects.add(refConfigPath);
+    const parsedRef = ts.getParsedCommandLineOfConfigFile(
+      refConfigPath,
+      undefined,
+      ts.sys as unknown as ts.ParseConfigFileHost
+    );
+    if (parsedRef) {
+      sourceFiles.push(...parsedRef.fileNames);
+      sourceFiles.push(
+        ...collectProjectReferenceSourceFiles(
+          parsedRef.projectReferences,
+          visitedProjects
+        )
+      );
+    }
+  }
+  return sourceFiles;
+}
+
 export class ReadonlyVisitor {
   public readonly key = '@nestjs/swagger';
   private readonly modelClassVisitor = new ModelClassVisitor();
   private readonly controllerClassVisitor = new ControllerClassVisitor();
+
+  /**
+   * Creates a TypeScript {@link ts.Program} from a `tsconfig.json` file path, correctly
+   * handling TypeScript project references (`references` field in tsconfig).
+   *
+   * When a project uses composite project references, calling `ts.createProgram` with
+   * the `projectReferences` option causes TypeScript to require pre-built output files
+   * (`.d.ts`) from each referenced project (error TS6305). This method avoids that by
+   * resolving all transitively referenced project source files and including them directly
+   * in the program's root file names, so no pre-built output is required.
+   *
+   * @param tsconfigPath Absolute path to the `tsconfig.json` file.
+   * @returns A `ts.Program` ready to be passed to {@link ReadonlyVisitor#visit}.
+   */
+  static createTsProgram(tsconfigPath: string): ts.Program {
+    let parseError: ts.Diagnostic | undefined;
+    const host: ts.ParseConfigFileHost = {
+      ...(ts.sys as unknown as ts.ParseConfigFileHost),
+      onUnRecoverableConfigFileDiagnostic(diagnostic: ts.Diagnostic) {
+        parseError = diagnostic;
+      }
+    };
+    const parsedCmd = ts.getParsedCommandLineOfConfigFile(
+      tsconfigPath,
+      undefined,
+      host
+    );
+    if (!parsedCmd || parseError) {
+      const message = parseError
+        ? ts.flattenDiagnosticMessageText(parseError.messageText, '\n')
+        : tsconfigPath;
+      throw new Error(`Failed to parse tsconfig at path: ${message}`);
+    }
+    const { options, fileNames, projectReferences } = parsedCmd;
+    const referencedSourceFiles =
+      collectProjectReferenceSourceFiles(projectReferences);
+    const rootNames = [...new Set([...fileNames, ...referencedSourceFiles])];
+    return ts.createProgram({ options, rootNames });
+  }
 
   get typeImports() {
     return {

--- a/test/plugin/fixtures/project-references/app/src/app.dto.ts
+++ b/test/plugin/fixtures/project-references/app/src/app.dto.ts
@@ -1,0 +1,6 @@
+import { SharedDto } from '../../libs/shared/src/shared.dto';
+
+export class AppDto {
+  shared: SharedDto;
+  title: string;
+}

--- a/test/plugin/fixtures/project-references/app/tsconfig.json
+++ b/test/plugin/fixtures/project-references/app/tsconfig.json
@@ -1,0 +1,16 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
+    "declaration": true,
+    "emitDecoratorMetadata": true,
+    "experimentalDecorators": true,
+    "target": "ES2021",
+    "outDir": "./dist",
+    "rootDir": "./src",
+    "skipLibCheck": true
+  },
+  "references": [
+    { "path": "../libs/shared" }
+  ],
+  "include": ["src/**/*"]
+}

--- a/test/plugin/fixtures/project-references/libs/shared/src/shared.dto.ts
+++ b/test/plugin/fixtures/project-references/libs/shared/src/shared.dto.ts
@@ -1,0 +1,4 @@
+export class SharedDto {
+  id: number;
+  name: string;
+}

--- a/test/plugin/fixtures/project-references/libs/shared/tsconfig.json
+++ b/test/plugin/fixtures/project-references/libs/shared/tsconfig.json
@@ -1,0 +1,14 @@
+{
+  "compilerOptions": {
+    "composite": true,
+    "module": "commonjs",
+    "declaration": true,
+    "emitDecoratorMetadata": true,
+    "experimentalDecorators": true,
+    "target": "ES2021",
+    "outDir": "./dist",
+    "rootDir": "./src",
+    "skipLibCheck": true
+  },
+  "include": ["src/**/*"]
+}

--- a/test/plugin/readonly-visitor-project-refs.spec.ts
+++ b/test/plugin/readonly-visitor-project-refs.spec.ts
@@ -1,0 +1,130 @@
+import { join } from 'path';
+import * as ts from 'typescript';
+import { ReadonlyVisitor } from '../../lib/plugin/visitors/readonly.visitor';
+
+const APP_TSCONFIG = join(
+  __dirname,
+  'fixtures',
+  'project-references',
+  'app',
+  'tsconfig.json'
+);
+
+describe('ReadonlyVisitor with TypeScript project references', () => {
+  describe('ReadonlyVisitor.createTsProgram', () => {
+    it('should create a program without TS6305 errors when referenced composite projects have not been built', () => {
+      // Verify no pre-built output exists for the referenced project so the test
+      // actually exercises the fix.
+      const sharedDistIndex = join(
+        __dirname,
+        'fixtures',
+        'project-references',
+        'libs',
+        'shared',
+        'dist',
+        'shared.dto.d.ts'
+      );
+      expect(ts.sys.fileExists(sharedDistIndex)).toBe(false);
+
+      // Before the fix, the common pattern of passing `projectReferences` to
+      // `ts.createProgram` caused TS6305 when referenced composite projects
+      // had not been pre-built. Verify the old pattern does produce TS6305 to
+      // confirm the fixture correctly reproduces the bug.
+      const parsedCmd = ts.getParsedCommandLineOfConfigFile(
+        APP_TSCONFIG,
+        undefined,
+        ts.sys as unknown as ts.ParseConfigFileHost
+      );
+      expect(parsedCmd).toBeDefined();
+      const { options, fileNames: rootNames, projectReferences } = parsedCmd!;
+      expect(projectReferences).toBeDefined();
+      expect(projectReferences!.length).toBeGreaterThan(0);
+
+      const buggyProgram = ts.createProgram({ options, rootNames, projectReferences });
+      const buggyDiags = ts.getPreEmitDiagnostics(buggyProgram);
+      const ts6305Errors = buggyDiags.filter((d) => d.code === 6305);
+      expect(ts6305Errors.length).toBeGreaterThan(0);
+
+      // The fix: use ReadonlyVisitor.createTsProgram which resolves project
+      // reference source files directly, avoiding the TS6305 error.
+      const program = ReadonlyVisitor.createTsProgram(APP_TSCONFIG);
+      const diagnostics = ts.getPreEmitDiagnostics(program);
+      const fixedTs6305Errors = diagnostics.filter((d) => d.code === 6305);
+
+      expect(fixedTs6305Errors.length).toBe(0);
+    });
+
+    it('should include source files from referenced projects in the program', () => {
+      const program = ReadonlyVisitor.createTsProgram(APP_TSCONFIG);
+
+      const sourceFileNames = program
+        .getSourceFiles()
+        .filter((sf) => !sf.isDeclarationFile)
+        .map((sf) => sf.fileName.replace(/\\/g, '/'));
+
+      // The app source file must be present
+      expect(
+        sourceFileNames.some((f) => f.includes('app.dto.ts'))
+      ).toBe(true);
+
+      // The referenced project's source file must also be present
+      expect(
+        sourceFileNames.some((f) => f.includes('shared.dto.ts'))
+      ).toBe(true);
+    });
+
+    it('should preserve type information from referenced project source files', () => {
+      const program = ReadonlyVisitor.createTsProgram(APP_TSCONFIG);
+      const checker = program.getTypeChecker();
+
+      const appSourceFile = program
+        .getSourceFiles()
+        .find((sf) => !sf.isDeclarationFile && sf.fileName.includes('app.dto.ts'));
+
+      expect(appSourceFile).toBeDefined();
+
+      let sharedPropertyType: string | undefined;
+      appSourceFile!.forEachChild((node) => {
+        if (ts.isClassDeclaration(node) && node.name?.text === 'AppDto') {
+          node.members.forEach((member) => {
+            if (
+              ts.isPropertyDeclaration(member) &&
+              ts.isIdentifier(member.name) &&
+              member.name.text === 'shared'
+            ) {
+              const type = checker.getTypeAtLocation(member);
+              sharedPropertyType = checker.typeToString(type);
+            }
+          });
+        }
+      });
+
+      // Type info from the referenced project should be resolvable
+      expect(sharedPropertyType).toBe('SharedDto');
+    });
+
+    it('should handle a tsconfig without project references', () => {
+      const plainTsconfig = join(
+        __dirname,
+        'fixtures',
+        'project',
+        'tsconfig.json'
+      );
+
+      // Should not throw for a tsconfig without references
+      expect(() => ReadonlyVisitor.createTsProgram(plainTsconfig)).not.toThrow();
+
+      const program = ReadonlyVisitor.createTsProgram(plainTsconfig);
+      expect(program).toBeDefined();
+    });
+
+    it('should throw when tsconfig path does not exist', () => {
+      const nonExistentPath = join(
+        __dirname,
+        'fixtures',
+        'non-existent-tsconfig.json'
+      );
+      expect(() => ReadonlyVisitor.createTsProgram(nonExistentPath)).toThrow();
+    });
+  });
+});

